### DISCUSSION TITLE
How to: trigger materializing an asset's partition upon upstream asset partition materialization

### DISCUSSION THREAD

**Discussion by clairelin135** [dagster-io/dagster#7306] [Q&A] [answered]
We have a partitioned asset X depends on three other partitioned assets. How can we trigger a run that materializes partition P of asset X whenever partition P has been materialized for the three upstream assets?

**Comment by clairelin135** [comment, marked as answer]
One possible way to do this is to define a sensor that checks for the existence of materializations for the upstream asset partitions:
```
from typing import Dict, Set

from dagster import (
    AssetGroup,
    AssetKey,
    AssetMaterialization,
    DagsterEventType,
    DailyPartitionsDefinition,
    EventRecordsFilter,
    Output,
    RunRequest,
    asset,
    job,
    op,
    repository,
    sensor,
)

daily_partitions_def = DailyPartitionsDefinition(start_date="2020-01-01")

@asset(partitions_def=daily_partitions_def)
def table_a(context):
    return 1

@asset(partitions_def=daily_partitions_def)
def table_b(context):
    return 1

@asset(partitions_def=daily_partitions_def)
def table_c(context):
    return 1

@asset(partitions_def=daily_partitions_def)
def dependent_asset(context):
    return 1

asset_group = AssetGroup.from_current_module()
trigger_dependent_asset = asset_group.build_job(name="trigger_dependent_asset", selection=['dependent_asset'])

@sensor(job=trigger_dependent_asset)
def multi_asset_sensor(context):
    partition_keys = daily_partitions_def.get_partitions_keys()
    last_partition_index = int(context.cursor) if context.cursor else -1
    curr_partition = partition_keys[last_partition_index + 1]

    asset_partition_materialized: Dict[AssetKey, bool] = {} # mapping of asset key to dictionary of materialization status by partition

    asset_keys = [AssetKey("table_a"), AssetKey("table_b"), AssetKey("table_c")]
    for asset_key in asset_keys:
        records = context.instance.get_event_records(
            EventRecordsFilter(
                event_type=DagsterEventType.ASSET_MATERIALIZATION,
                asset_key=asset_key,
                asset_partitions=[curr_partition],
            )
        )
        asset_partition_materialized[asset_key] = True if len(records) else False # materialization record exists for partition

    if asset_partition_materialized[AssetKey("table_a")] and asset_partition_materialized[AssetKey("table_b")] and asset_partition_materialized[AssetKey("table_c")]:
        yield trigger_dependent_asset.run_request_for_partition(partition_key=curr_partition, run_key=None)
        context.update_cursor(str(last_partition_index + 1))


@repository
def my_repo():
    return [
        asset_group, trigger_dependent_asset, multi_asset_sensor
    ]
```
In the above example, because the assets are time-window partitioned, we can use the `SensorEvaluationContext.cursor` object to track the latest executed partition. Then, we can search only for the materializations of the current partition.

**Reply by geoHeil** [reply]
Why is the `run_key=None`? Why not the `partition_key`? Isn`t this usually required for idempotence?

**Reply by clairelin135** [reply]
Hi geoHeil. Sensors maintain that at most one run is created for a given `run_key`, so providing a run key and triggering a run from the sensor would result in skipping future run requests with the same run key.

The partition key is used to specify which partition the job should be executed with, which can be accessed from the op execution context via `context.partition_key`.

**Reply by geoHeil** [reply]
Understood. But oftentimes, this will translate the same thing right?

**Reply by clairelin135** [reply]
Yep--if you provided the `run_key=partition_key` this would translate to the same thing.

**Reply by geoHeil** [reply]
Can you explain to me how to modify this code example to perform the fan-in like you are explaining here, but: It should only trigger the (unpartitioned) downstream asset once all partitions for a specific date (of the upstream assets) have arrived.

**Reply by geoHeil** [reply]
I have prepared my take on this: https://github.com/geoHeil/dagster-ssh-demo/blob/master/SSH_DEMO/sensors/sftp_sensor_asset_real.py#L206 perhaps you can comment if this is the desired way to achieve this. At least to me, it looks like this should be correct.

**Reply by clairelin135** [reply]
Hey @geoHeil, yes that looks correct.

**Reply by simonvanderveldt** [reply]
@clairelin135 How would you rate the use of `context.instance.get_event_records()` with a filter on the desired partition vs using `context.instance.get_asset_records()` and checking the partition value in the returned `AssetRecord`?
I guess `get_event_records()` is the better approach since that checks for a specific partition whereas `get_asset_records` would only allow checking against the latest partition.

**Reply by geoHeil** [reply]
See the ongoing discussion here: https://dagster.slack.com/archives/C01U954MEER/p1650091358075689?thread_ts=1649909251.959589&cid=C01U954MEER also for me some topics are not really clear yet here.

**Reply by clairelin135** [reply]
@simonvanderveldt Yes, exactly. `get_asset_records` would only return the latest materialization for the given asset. `get_event_records` is what you would be looking for in this case.

**Reply by simonvanderveldt** [reply]
@clairelin135 All clear, thanks for the confirmation!
